core/node/vm_runner/src/impls/bwip.rs

```diff
@@ -0,0 +1,447 @@
+/// intended.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerTasks {
+    /// Task that synchronizes storage with new available batches.
+    pub loader_task: StorageSyncTask<BasicWitnessInputProducerIo>,
+    /// Task that handles output from processed batches.
+    pub output_handler_factory_task:
+        ConcurrentOutputHandlerFactoryTask<BasicWitnessInputProducerIo>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BasicWitnessInputProducerIo {
+    first_processed_batch: L1BatchNumber,
+    window_size: u32,
+}
+
+#[async_trait]
+impl VmRunnerIo for BasicWitnessInputProducerIo {
+    fn name(&self) -> &'static str {
+        "basic_witness_input_producer"
+    }
+
+    async fn latest_processed_batch(
+        &self,
+        conn: &mut Connection<'_, Core>,
+    ) -> anyhow::Result<L1BatchNumber> {
+        Ok(conn
+            .vm_runner_dal()
+            .get_bwip_latest_processed_batch(self.first_processed_batch)
+            .await?)
+    }
+
+    async fn last_ready_to_be_loaded_batch(
+        &self,
+        conn: &mut Connection<'_, Core>,
+    ) -> anyhow::Result<L1BatchNumber> {
+        Ok(conn
+            .vm_runner_dal()
+            .get_bwip_last_ready_batch(self.first_processed_batch, self.window_size)
+            .await?)
+    }
+
+    async fn mark_l1_batch_as_completed(
+        &self,
+        conn: &mut Connection<'_, Core>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<()> {
+        Ok(conn
+            .vm_runner_dal()
+            .mark_bwip_batch_as_completed(l1_batch_number)
+            .await?)
+    }
+}
+
+#[derive(Debug)]
+struct BasicWitnessInputProducerOutputHandler {
+    pool: ConnectionPool<Core>,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+#[async_trait]
+impl StateKeeperOutputHandler for BasicWitnessInputProducerOutputHandler {
+    async fn handle_l2_block(&mut self, _updates_manager: &UpdatesManager) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    async fn handle_l1_batch(
+        &mut self,
+        updates_manager: Arc<UpdatesManager>,
+    ) -> anyhow::Result<()> {
+        let l1_batch_number = updates_manager.l1_batch.number;
+
+        tracing::info!(
+            "Started saving VM run data for L1 batch {}",
+            l1_batch_number
+        );
+
+        let db_result =
+            get_database_witness_input_data(&mut self.pool.connection().await?, l1_batch_number)
+                .await;
+        let mut result = get_updates_manager_witness_input_data(
+            &mut self.pool.connection().await?,
+            updates_manager.clone(),
+        )
+        .await;
+
+        compare_witness_input_data(&db_result, &result);
+
+        let previous_batch_with_metadata = self
+            .pool
+            .connection()
+            .await?
+            .blocks_dal()
+            .get_l1_batch_metadata(L1BatchNumber(l1_batch_number.checked_sub(1).unwrap()))
+            .await
+            .unwrap()
+            .unwrap();
```

Comment:
You have a few unwraps in this method, any reason not to propagate them upstream?